{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c222b208",
   "metadata": {},
   "source": [
    "# Mastering DNA/Protein Sequence Analysis in R: A Step-by-Step Tutorial for Beginners\n",
    "\n",
    "Find more tutorials about Computer Vision, Microscopy, Biology and Data Science [here](https://medium.com/@microbioscopicdata)\n",
    "\n",
    "![SegmentLocal](Animated_logo_smaller_faster.gif \"segment\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8e92df9a",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\"> Sequences are the fundamentals tools for biologists and bioinformaticians playing a crucial role in various domains of research. Sequence analysis has made significant contributions to biological studies and remains a fundamental task with broad applications in genetics, genomics, text mining, and time series analysis 1. Whether you are a biologist, a data scientist, or simply someone curious about exploring patterns and structures in sequences, this comprehensive tutorial aims to provide you with a solid foundation in DNA/protein sequence analysis using R. We will discuss how to read and store DNA/protein sequences, manipulate and transform sequences, find motifs, and calculate sequence statistics.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bb10c3b",
   "metadata": {},
   "source": [
    "## Install necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7ad17dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install the seqinr package if not already installed, uncomment the line below\n",
    "# install.packages(\"seqinr\") \n",
    "\n",
    "# Install the APE package if not already installed, uncomment the line below\n",
    "# install.packages(\"ape\")\n",
    "\n",
    "# Install the Biostrings package using Bioconductor if not already installed, uncomment the line below\n",
    "# if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#     install.packages(\"BiocManager\")\n",
    "# BiocManager::install(\"Biostrings\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85c98f3a",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "88c9b2d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(seqinr);\n",
    "library(ape);\n",
    "library(rentrez);\n",
    "library(Biostrings);"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e8469a7b",
   "metadata": {},
   "source": [
    "## Creating, downloading, reading and saving sequences in R\n",
    "<div style=\"text-align: justify\">Biological sequences are most often nucleotide or protein sequences usually represented as simple strings of characters. In R, DNA or protein sequences are typically represented as character vectors, which are simple strings of characters.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "2b4375bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ATGCTGACTGCTAGCGATCGTAGC\"\n"
     ]
    }
   ],
   "source": [
    "# DNA sequence as character vector\n",
    "dna_seq <- \"ATGCTGACTGCTAGCGATCGTAGC\"\n",
    "print(dna_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7265251",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\">In R we can retrieve sequences data from NCBI (National Center for Biotechnology Information) using the rentrez package. In this tutorial, we will demonstrate how to download (and save in FASTA format, see below) gene sequences associated with plasma-membrane-associated protein complexes in the model filamentous fungus <i>Aspergillus nidulans</i>, such as <i>pilA</i> and <i>pilB</i> 3,4. \n",
    "The <i>pilA</i> (and <i>pilB</i>) genes encode for BAR-domain-containing proteins that localize to eisosomes, which are intriguing cellular structures with multifaceted functions. Eisosomes are involved in crucial cellular processes, including plasma membrane organization, cell wall synthesis and morphogenesis, sphingolipid homeostasis, and stress response 5.</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "3efdb7fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch the sequence from NCBI - pilA\n",
    "pilA <- rentrez::entrez_fetch(db = \"nucleotide\", \n",
    "                          id = \"XM_657729.2\",\n",
    "                          rettype = \"fasta\")\n",
    "# Save the sequence pilA in a FASTA file\n",
    "write(pilA, file=\"pilA.fasta\")\n",
    "\n",
    "# Fetch the sequence from NCBI - pilB\n",
    "pilB <- rentrez::entrez_fetch(db = \"nucleotide\", \n",
    "                          id = \"XM_050611370.1\",\n",
    "                          rettype = \"fasta\")\n",
    "# Save the sequence pilB in a FASTA file\n",
    "write(pilB, file=\"pilB.fasta\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5c2c25b0",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\">Sequences are commonly saved and shared in the FASTA format, which is a simple and widely used format for storing biological sequences such as DNA or protein sequences2. In FASTA files each record (FASTA files can store one or multiple sequences/records) begin with a description line starting with a greater-than sign > character, followed on the next line by the sequences.</div>\n",
    "\n",
    "`>XM_657729.2 Aspergillus nidulans FGSC A4 lipid-binding protein pilA (ANIA_05217), mRNA\n",
    "TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "b5579a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the file path\n",
    "fasta_file <- \"pilA.fasta\"\n",
    "\n",
    "# Read the FASTA file using read.fasta()\n",
    "pilA <- read.fasta(file = fasta_file)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9b7da3e7",
   "metadata": {},
   "source": [
    "## Convert and clean FASTA sequences\n",
    "\n",
    "To manipulate sequences and address various biological questions, it's essential to convert FASTA sequences into a suitable data structure in R 6.  . During the conversion process, there are several elements we need to remove or modify:  \n",
    "\n",
    "•\tRemove the identifier line, which typically starts with the \">\" symbol and contains information about the sequence, such as the accession number and description.  \n",
    "•\tEliminate all line breaks (\"\\n\") that appear in the file, as they are not required for further analysis.  \n",
    "\n",
    "\n",
    "•\t(optional) Organize each nucleotide of the sequence into its individual position within a vector, allowing for efficient data manipulation. \n",
    "\n",
    "By performing these steps, we can obtain a clean and structured representation of the sequence data that is ready for further analysis and exploration in R."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "22932eea",
   "metadata": {},
   "source": [
    "### Remove the identifier line (metadata header). \n",
    "\n",
    "To remove  the identifier line we are going to use regular expressions. Regular expressions allow you to define patterns to find and replace specific characters or sequences.We are going to use gsub() function ito remove the pattern ^>.*?\\n that matches the identifier line starting with >, followed by any characters .*?, and the newline character \\n. \n",
    "\n",
    "^>: The ^ symbol asserts the beginning of a line, and > matches the greater-than symbol at the start of the metadata line. So, ^> matches any line that starts with >, which is typically the marker for the metadata lines in a FASTA file.  \n",
    "\n",
    ".*?: The . matches any character except a newline. The * quantifier means zero or more occurrences of the previous character or group. The ? makes the * quantifier lazy, which means it will match as few characters as possible.   \n",
    "This combination .*? matches any character lazily until the first occurrence of the following pattern.  \n",
    "\n",
    "\\n: The \\n represents the newline character. It indicates the end of a line.\n",
    "Therefore, the pattern \"^>.*?\\n\" matches the metadata line from the beginning of the line until the first newline character."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "d88f79ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'&gt;XM_657729.2 Aspergillus nidulans FGSC A4 lipid-binding protein pilA (ANIA_05217), mRNA\\nTTTCGAACAGAAATCTGCTGTCCATCCACCAGT'"
      ],
      "text/latex": [
       "'>XM\\_657729.2 Aspergillus nidulans FGSC A4 lipid-binding protein pilA (ANIA\\_05217), mRNA\\textbackslash{}nTTTCGAACAGAAATCTGCTGTCCATCCACCAGT'"
      ],
      "text/markdown": [
       "'&gt;XM_657729.2 Aspergillus nidulans FGSC A4 lipid-binding protein pilA (ANIA_05217), mRNA\\nTTTCGAACAGAAATCTGCTGTCCATCCACCAGT'"
      ],
      "text/plain": [
       "[1] \">XM_657729.2 Aspergillus nidulans FGSC A4 lipid-binding protein pilA (ANIA_05217), mRNA\\nTTTCGAACAGAAATCTGCTGTCCATCCACCAGT\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC\\nAGATTGCTGGTCCTTATTTGAGAGTCACTAATACTGCTTGTGAGCTCGTG'"
      ],
      "text/latex": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC\\textbackslash{}nAGATTGCTGGTCCTTATTTGAGAGTCACTAATACTGCTTGTGAGCTCGTG'"
      ],
      "text/markdown": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC\\nAGATTGCTGGTCCTTATTTGAGAGTCACTAATACTGCTTGTGAGCTCGTG'"
      ],
      "text/plain": [
       "[1] \"TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC\\nAGATTGCTGGTCCTTATTTGAGAGTCACTAATACTGCTTGTGAGCTCGTG\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Fetch the sequence from NCBI - pilA\n",
    "pilA <- rentrez::entrez_fetch(db = \"nucleotide\", \n",
    "                          id = \"XM_657729.2\",\n",
    "                          rettype = \"fasta\")\n",
    "pilA_cleaned <- gsub(\"^>.*?\\n\", \"\", replacement = \"\",x = pilA)\n",
    "# Print the original sequence downloaded from NCBI (trim the sequence length to 120 characters for display)\n",
    "strtrim(pilA,120)\n",
    "# Print the cleaned sequence (trim the sequence length to 120 characters for display)\n",
    "strtrim(pilA_cleaned,120)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8d3f493a",
   "metadata": {},
   "source": [
    "### Eliminate all line breaks (\"\\n\")\n",
    "\n",
    "<div style=\"text-align: justify\">To remove unwanted characters, including the newline character (\\n), you can use regular expressions with the gsub() function in R. In this case, we want to remove the \\n characters from the sequence. Since the backslash () is a special character in regular expressions, we need to escape it by using two backslashes (\\).</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "ca68ecfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC\\nAGATT'"
      ],
      "text/latex": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC\\textbackslash{}nAGATT'"
      ],
      "text/markdown": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC\\nAGATT'"
      ],
      "text/plain": [
       "[1] \"TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCC\\nAGATT\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCCAGATT'"
      ],
      "text/latex": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCCAGATT'"
      ],
      "text/markdown": [
       "'TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCCAGATT'"
      ],
      "text/plain": [
       "[1] \"TTTCGAACAGAAATCTGCTGTCCATCCACCAGTCAAGACGCGAATTGACTGTCGCGGCCCCAGTCGTCCCAGATT\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove all line breaks (\"\\n\") using gsub function\n",
    "pilA_vector <- gsub(pattern = \"\\\\n\", replacement = \"\",x = pilA_cleaned)\n",
    "\n",
    "# Print the sequences (trim the sequence length to 120 characters for display)\n",
    "strtrim(pilA_cleaned,75)\n",
    "strtrim(pilA_vector,75)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a7c138dc",
   "metadata": {},
   "source": [
    "### Organize each nucleotide of the sequence into its individual position within a vector\n",
    "\n",
    "At last we need to split a continuous string of letters into a vector where each letter is separated, you can utilize the `str_split()` function from the stringr package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "323888bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'T'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'G'</li><li>'A'</li><li>'A'</li><li>'C'</li><li>'A'</li><li>'G'</li><li>'A'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'A'</li><li>'G'</li><li>'A'</li><li>'C'</li><li>'G'</li><li>'C'</li><li>'G'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'G'</li><li>'C'</li><li>'G'</li><li>'G'</li><li>'C'</li><li>'C'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'G'</li><li>'A'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'C'</li><li>'G'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'G'</li><li>'G'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'G'</li><li>'C'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'A'</li><li>'C'</li><li>'C'</li><li>'G'</li><li>'A'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'G'</li><li>'A'</li><li>'C'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'C'</li><li>'C'</li><li>'G'</li><li>'A'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'C'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'C'</li><li>'G'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'C'</li><li>'T'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'A'</li><li>'C'</li><li>⋯</li><li>'A'</li><li>'C'</li><li>'C'</li><li>'G'</li><li>'G'</li><li>'C'</li><li>'A'</li><li>'A'</li><li>'A'</li><li>'A'</li><li>'C'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'A'</li><li>'C'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'G'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'G'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'T'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'T'</li><li>'A'</li><li>'C'</li><li>'G'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'T'</li><li>'T'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'C'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'G'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'C'</li><li>'G'</li><li>'G'</li><li>'T'</li><li>'G'</li><li>'T'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'C'</li><li>'G'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'G'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'C'</li><li>'C'</li><li>'C'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'G'</li><li>'G'</li><li>'T'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'A'</li><li>'T'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'A'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'T'</li><li>'G'</li><li>'G'</li><li>'C'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'A'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'A'</li><li>'A'</li><li>'A'</li><li>'A'</li><li>'C'</li><li>'G'</li><li>'C'</li><li>'T'</li><li>'T'</li><li>'C'</li><li>'C'</li><li>'C'</li><li>'C'</li><li>'A'</li><li>'A'</li><li>'T'</li><li>'T'</li><li>'G'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item ⋯\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'C'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'C'\n",
       "\\item 'A'\n",
       "\\item 'A'\n",
       "\\item 'T'\n",
       "\\item 'T'\n",
       "\\item 'G'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'T'\n",
       "2. 'T'\n",
       "3. 'T'\n",
       "4. 'C'\n",
       "5. 'G'\n",
       "6. 'A'\n",
       "7. 'A'\n",
       "8. 'C'\n",
       "9. 'A'\n",
       "10. 'G'\n",
       "11. 'A'\n",
       "12. 'A'\n",
       "13. 'A'\n",
       "14. 'T'\n",
       "15. 'C'\n",
       "16. 'T'\n",
       "17. 'G'\n",
       "18. 'C'\n",
       "19. 'T'\n",
       "20. 'G'\n",
       "21. 'T'\n",
       "22. 'C'\n",
       "23. 'C'\n",
       "24. 'A'\n",
       "25. 'T'\n",
       "26. 'C'\n",
       "27. 'C'\n",
       "28. 'A'\n",
       "29. 'C'\n",
       "30. 'C'\n",
       "31. 'A'\n",
       "32. 'G'\n",
       "33. 'T'\n",
       "34. 'C'\n",
       "35. 'A'\n",
       "36. 'A'\n",
       "37. 'G'\n",
       "38. 'A'\n",
       "39. 'C'\n",
       "40. 'G'\n",
       "41. 'C'\n",
       "42. 'G'\n",
       "43. 'A'\n",
       "44. 'A'\n",
       "45. 'T'\n",
       "46. 'T'\n",
       "47. 'G'\n",
       "48. 'A'\n",
       "49. 'C'\n",
       "50. 'T'\n",
       "51. 'G'\n",
       "52. 'T'\n",
       "53. 'C'\n",
       "54. 'G'\n",
       "55. 'C'\n",
       "56. 'G'\n",
       "57. 'G'\n",
       "58. 'C'\n",
       "59. 'C'\n",
       "60. 'C'\n",
       "61. 'C'\n",
       "62. 'A'\n",
       "63. 'G'\n",
       "64. 'T'\n",
       "65. 'C'\n",
       "66. 'G'\n",
       "67. 'T'\n",
       "68. 'C'\n",
       "69. 'C'\n",
       "70. 'C'\n",
       "71. 'A'\n",
       "72. 'G'\n",
       "73. 'A'\n",
       "74. 'T'\n",
       "75. 'T'\n",
       "76. 'G'\n",
       "77. 'C'\n",
       "78. 'T'\n",
       "79. 'G'\n",
       "80. 'G'\n",
       "81. 'T'\n",
       "82. 'C'\n",
       "83. 'C'\n",
       "84. 'T'\n",
       "85. 'T'\n",
       "86. 'A'\n",
       "87. 'T'\n",
       "88. 'T'\n",
       "89. 'T'\n",
       "90. 'G'\n",
       "91. 'A'\n",
       "92. 'G'\n",
       "93. 'A'\n",
       "94. 'G'\n",
       "95. 'T'\n",
       "96. 'C'\n",
       "97. 'A'\n",
       "98. 'C'\n",
       "99. 'T'\n",
       "100. 'A'\n",
       "101. 'A'\n",
       "102. 'T'\n",
       "103. 'A'\n",
       "104. 'C'\n",
       "105. 'T'\n",
       "106. 'G'\n",
       "107. 'C'\n",
       "108. 'T'\n",
       "109. 'T'\n",
       "110. 'G'\n",
       "111. 'T'\n",
       "112. 'G'\n",
       "113. 'A'\n",
       "114. 'G'\n",
       "115. 'C'\n",
       "116. 'T'\n",
       "117. 'C'\n",
       "118. 'G'\n",
       "119. 'T'\n",
       "120. 'G'\n",
       "121. 'A'\n",
       "122. 'C'\n",
       "123. 'T'\n",
       "124. 'T'\n",
       "125. 'C'\n",
       "126. 'C'\n",
       "127. 'A'\n",
       "128. 'T'\n",
       "129. 'T'\n",
       "130. 'T'\n",
       "131. 'C'\n",
       "132. 'C'\n",
       "133. 'G'\n",
       "134. 'G'\n",
       "135. 'T'\n",
       "136. 'T'\n",
       "137. 'A'\n",
       "138. 'G'\n",
       "139. 'C'\n",
       "140. 'G'\n",
       "141. 'T'\n",
       "142. 'C'\n",
       "143. 'C'\n",
       "144. 'A'\n",
       "145. 'A'\n",
       "146. 'C'\n",
       "147. 'C'\n",
       "148. 'G'\n",
       "149. 'A'\n",
       "150. 'G'\n",
       "151. 'T'\n",
       "152. 'C'\n",
       "153. 'T'\n",
       "154. 'C'\n",
       "155. 'A'\n",
       "156. 'G'\n",
       "157. 'A'\n",
       "158. 'C'\n",
       "159. 'G'\n",
       "160. 'T'\n",
       "161. 'C'\n",
       "162. 'A'\n",
       "163. 'C'\n",
       "164. 'C'\n",
       "165. 'G'\n",
       "166. 'A'\n",
       "167. 'C'\n",
       "168. 'C'\n",
       "169. 'A'\n",
       "170. 'T'\n",
       "171. 'T'\n",
       "172. 'T'\n",
       "173. 'C'\n",
       "174. 'C'\n",
       "175. 'T'\n",
       "176. 'C'\n",
       "177. 'C'\n",
       "178. 'A'\n",
       "179. 'C'\n",
       "180. 'T'\n",
       "181. 'C'\n",
       "182. 'G'\n",
       "183. 'A'\n",
       "184. 'T'\n",
       "185. 'C'\n",
       "186. 'G'\n",
       "187. 'G'\n",
       "188. 'C'\n",
       "189. 'T'\n",
       "190. 'C'\n",
       "191. 'T'\n",
       "192. 'C'\n",
       "193. 'T'\n",
       "194. 'G'\n",
       "195. 'C'\n",
       "196. 'T'\n",
       "197. 'G'\n",
       "198. 'A'\n",
       "199. 'A'\n",
       "200. 'C'\n",
       "201. ⋯\n",
       "202. 'A'\n",
       "203. 'C'\n",
       "204. 'C'\n",
       "205. 'G'\n",
       "206. 'G'\n",
       "207. 'C'\n",
       "208. 'A'\n",
       "209. 'A'\n",
       "210. 'A'\n",
       "211. 'A'\n",
       "212. 'C'\n",
       "213. 'G'\n",
       "214. 'A'\n",
       "215. 'T'\n",
       "216. 'A'\n",
       "217. 'C'\n",
       "218. 'A'\n",
       "219. 'T'\n",
       "220. 'T'\n",
       "221. 'G'\n",
       "222. 'G'\n",
       "223. 'T'\n",
       "224. 'T'\n",
       "225. 'A'\n",
       "226. 'A'\n",
       "227. 'T'\n",
       "228. 'C'\n",
       "229. 'A'\n",
       "230. 'C'\n",
       "231. 'T'\n",
       "232. 'T'\n",
       "233. 'T'\n",
       "234. 'G'\n",
       "235. 'A'\n",
       "236. 'T'\n",
       "237. 'G'\n",
       "238. 'C'\n",
       "239. 'T'\n",
       "240. 'C'\n",
       "241. 'T'\n",
       "242. 'T'\n",
       "243. 'G'\n",
       "244. 'G'\n",
       "245. 'A'\n",
       "246. 'C'\n",
       "247. 'T'\n",
       "248. 'C'\n",
       "249. 'C'\n",
       "250. 'A'\n",
       "251. 'T'\n",
       "252. 'A'\n",
       "253. 'A'\n",
       "254. 'T'\n",
       "255. 'T'\n",
       "256. 'A'\n",
       "257. 'T'\n",
       "258. 'T'\n",
       "259. 'G'\n",
       "260. 'T'\n",
       "261. 'A'\n",
       "262. 'C'\n",
       "263. 'G'\n",
       "264. 'A'\n",
       "265. 'A'\n",
       "266. 'T'\n",
       "267. 'G'\n",
       "268. 'A'\n",
       "269. 'C'\n",
       "270. 'T'\n",
       "271. 'G'\n",
       "272. 'T'\n",
       "273. 'T'\n",
       "274. 'T'\n",
       "275. 'T'\n",
       "276. 'C'\n",
       "277. 'C'\n",
       "278. 'C'\n",
       "279. 'G'\n",
       "280. 'C'\n",
       "281. 'T'\n",
       "282. 'T'\n",
       "283. 'G'\n",
       "284. 'A'\n",
       "285. 'T'\n",
       "286. 'C'\n",
       "287. 'T'\n",
       "288. 'T'\n",
       "289. 'C'\n",
       "290. 'A'\n",
       "291. 'G'\n",
       "292. 'G'\n",
       "293. 'A'\n",
       "294. 'T'\n",
       "295. 'T'\n",
       "296. 'C'\n",
       "297. 'A'\n",
       "298. 'G'\n",
       "299. 'A'\n",
       "300. 'T'\n",
       "301. 'C'\n",
       "302. 'G'\n",
       "303. 'G'\n",
       "304. 'T'\n",
       "305. 'G'\n",
       "306. 'T'\n",
       "307. 'C'\n",
       "308. 'T'\n",
       "309. 'G'\n",
       "310. 'C'\n",
       "311. 'C'\n",
       "312. 'A'\n",
       "313. 'C'\n",
       "314. 'G'\n",
       "315. 'T'\n",
       "316. 'T'\n",
       "317. 'A'\n",
       "318. 'T'\n",
       "319. 'G'\n",
       "320. 'A'\n",
       "321. 'T'\n",
       "322. 'G'\n",
       "323. 'T'\n",
       "324. 'T'\n",
       "325. 'A'\n",
       "326. 'C'\n",
       "327. 'T'\n",
       "328. 'T'\n",
       "329. 'A'\n",
       "330. 'T'\n",
       "331. 'T'\n",
       "332. 'A'\n",
       "333. 'C'\n",
       "334. 'C'\n",
       "335. 'C'\n",
       "336. 'C'\n",
       "337. 'T'\n",
       "338. 'T'\n",
       "339. 'C'\n",
       "340. 'G'\n",
       "341. 'G'\n",
       "342. 'T'\n",
       "343. 'T'\n",
       "344. 'T'\n",
       "345. 'G'\n",
       "346. 'A'\n",
       "347. 'T'\n",
       "348. 'A'\n",
       "349. 'A'\n",
       "350. 'T'\n",
       "351. 'G'\n",
       "352. 'C'\n",
       "353. 'T'\n",
       "354. 'T'\n",
       "355. 'A'\n",
       "356. 'C'\n",
       "357. 'T'\n",
       "358. 'T'\n",
       "359. 'G'\n",
       "360. 'C'\n",
       "361. 'T'\n",
       "362. 'T'\n",
       "363. 'C'\n",
       "364. 'T'\n",
       "365. 'G'\n",
       "366. 'G'\n",
       "367. 'C'\n",
       "368. 'A'\n",
       "369. 'A'\n",
       "370. 'T'\n",
       "371. 'A'\n",
       "372. 'A'\n",
       "373. 'T'\n",
       "374. 'C'\n",
       "375. 'A'\n",
       "376. 'A'\n",
       "377. 'C'\n",
       "378. 'C'\n",
       "379. 'A'\n",
       "380. 'A'\n",
       "381. 'T'\n",
       "382. 'T'\n",
       "383. 'C'\n",
       "384. 'A'\n",
       "385. 'A'\n",
       "386. 'A'\n",
       "387. 'A'\n",
       "388. 'C'\n",
       "389. 'G'\n",
       "390. 'C'\n",
       "391. 'T'\n",
       "392. 'T'\n",
       "393. 'C'\n",
       "394. 'C'\n",
       "395. 'C'\n",
       "396. 'C'\n",
       "397. 'A'\n",
       "398. 'A'\n",
       "399. 'T'\n",
       "400. 'T'\n",
       "401. 'G'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "   [1] \"T\" \"T\" \"T\" \"C\" \"G\" \"A\" \"A\" \"C\" \"A\" \"G\" \"A\" \"A\" \"A\" \"T\" \"C\" \"T\" \"G\" \"C\"\n",
       "  [19] \"T\" \"G\" \"T\" \"C\" \"C\" \"A\" \"T\" \"C\" \"C\" \"A\" \"C\" \"C\" \"A\" \"G\" \"T\" \"C\" \"A\" \"A\"\n",
       "  [37] \"G\" \"A\" \"C\" \"G\" \"C\" \"G\" \"A\" \"A\" \"T\" \"T\" \"G\" \"A\" \"C\" \"T\" \"G\" \"T\" \"C\" \"G\"\n",
       "  [55] \"C\" \"G\" \"G\" \"C\" \"C\" \"C\" \"C\" \"A\" \"G\" \"T\" \"C\" \"G\" \"T\" \"C\" \"C\" \"C\" \"A\" \"G\"\n",
       "  [73] \"A\" \"T\" \"T\" \"G\" \"C\" \"T\" \"G\" \"G\" \"T\" \"C\" \"C\" \"T\" \"T\" \"A\" \"T\" \"T\" \"T\" \"G\"\n",
       "  [91] \"A\" \"G\" \"A\" \"G\" \"T\" \"C\" \"A\" \"C\" \"T\" \"A\" \"A\" \"T\" \"A\" \"C\" \"T\" \"G\" \"C\" \"T\"\n",
       " [109] \"T\" \"G\" \"T\" \"G\" \"A\" \"G\" \"C\" \"T\" \"C\" \"G\" \"T\" \"G\" \"A\" \"C\" \"T\" \"T\" \"C\" \"C\"\n",
       " [127] \"A\" \"T\" \"T\" \"T\" \"C\" \"C\" \"G\" \"G\" \"T\" \"T\" \"A\" \"G\" \"C\" \"G\" \"T\" \"C\" \"C\" \"A\"\n",
       " [145] \"A\" \"C\" \"C\" \"G\" \"A\" \"G\" \"T\" \"C\" \"T\" \"C\" \"A\" \"G\" \"A\" \"C\" \"G\" \"T\" \"C\" \"A\"\n",
       " [163] \"C\" \"C\" \"G\" \"A\" \"C\" \"C\" \"A\" \"T\" \"T\" \"T\" \"C\" \"C\" \"T\" \"C\" \"C\" \"A\" \"C\" \"T\"\n",
       " [181] \"C\" \"G\" \"A\" \"T\" \"C\" \"G\" \"G\" \"C\" \"T\" \"C\" \"T\" \"C\" \"T\" \"G\" \"C\" \"T\" \"G\" \"A\"\n",
       " [199] \"A\" \"C\" \"T\" \"G\" \"G\" \"A\" \"T\" \"T\" \"C\" \"T\" \"C\" \"C\" \"G\" \"G\" \"G\" \"G\" \"A\" \"A\"\n",
       " [217] \"A\" \"G\" \"A\" \"T\" \"C\" \"T\" \"C\" \"C\" \"T\" \"A\" \"G\" \"G\" \"T\" \"C\" \"A\" \"T\" \"C\" \"A\"\n",
       " [235] \"A\" \"C\" \"T\" \"T\" \"T\" \"A\" \"T\" \"T\" \"G\" \"C\" \"A\" \"T\" \"T\" \"C\" \"T\" \"C\" \"C\" \"T\"\n",
       " [253] \"C\" \"T\" \"C\" \"G\" \"T\" \"C\" \"G\" \"A\" \"G\" \"C\" \"T\" \"A\" \"C\" \"C\" \"G\" \"A\" \"T\" \"T\"\n",
       " [271] \"G\" \"C\" \"G\" \"G\" \"T\" \"T\" \"C\" \"T\" \"C\" \"C\" \"G\" \"G\" \"C\" \"C\" \"G\" \"T\" \"T\" \"C\"\n",
       " [289] \"C\" \"G\" \"C\" \"C\" \"G\" \"T\" \"T\" \"T\" \"C\" \"C\" \"G\" \"T\" \"G\" \"T\" \"G\" \"A\" \"C\" \"C\"\n",
       " [307] \"T\" \"C\" \"C\" \"T\" \"T\" \"A\" \"T\" \"A\" \"C\" \"C\" \"A\" \"C\" \"A\" \"C\" \"C\" \"A\" \"C\" \"G\"\n",
       " [325] \"T\" \"C\" \"A\" \"C\" \"T\" \"C\" \"G\" \"T\" \"A\" \"C\" \"G\" \"C\" \"A\" \"A\" \"A\" \"G\" \"C\" \"G\"\n",
       " [343] \"G\" \"C\" \"C\" \"T\" \"G\" \"A\" \"G\" \"G\" \"A\" \"A\" \"C\" \"T\" \"C\" \"A\" \"T\" \"T\" \"C\" \"C\"\n",
       " [361] \"G\" \"T\" \"C\" \"C\" \"A\" \"A\" \"T\" \"C\" \"C\" \"G\" \"G\" \"C\" \"T\" \"C\" \"A\" \"A\" \"T\" \"T\"\n",
       " [379] \"A\" \"G\" \"T\" \"A\" \"G\" \"A\" \"A\" \"T\" \"C\" \"T\" \"T\" \"C\" \"T\" \"T\" \"C\" \"T\" \"C\" \"C\"\n",
       " [397] \"T\" \"T\" \"C\" \"T\" \"C\" \"G\" \"T\" \"C\" \"G\" \"C\" \"C\" \"C\" \"A\" \"C\" \"A\" \"A\" \"T\" \"C\"\n",
       " [415] \"T\" \"C\" \"T\" \"G\" \"T\" \"T\" \"C\" \"T\" \"T\" \"T\" \"G\" \"C\" \"C\" \"G\" \"A\" \"T\" \"C\" \"T\"\n",
       " [433] \"C\" \"G\" \"T\" \"A\" \"C\" \"G\" \"C\" \"A\" \"G\" \"A\" \"A\" \"T\" \"A\" \"T\" \"C\" \"G\" \"G\" \"C\"\n",
       " [451] \"T\" \"G\" \"G\" \"T\" \"C\" \"T\" \"C\" \"T\" \"T\" \"C\" \"A\" \"C\" \"A\" \"A\" \"T\" \"G\" \"C\" \"A\"\n",
       " [469] \"T\" \"C\" \"G\" \"T\" \"A\" \"C\" \"A\" \"T\" \"A\" \"C\" \"T\" \"C\" \"T\" \"A\" \"T\" \"G\" \"C\" \"G\"\n",
       " [487] \"C\" \"C\" \"A\" \"G\" \"T\" \"C\" \"G\" \"C\" \"G\" \"G\" \"C\" \"T\" \"G\" \"C\" \"C\" \"G\" \"A\" \"C\"\n",
       " [505] \"C\" \"G\" \"C\" \"C\" \"T\" \"C\" \"G\" \"C\" \"A\" \"G\" \"A\" \"T\" \"C\" \"G\" \"A\" \"G\" \"A\" \"A\"\n",
       " [523] \"C\" \"C\" \"C\" \"C\" \"C\" \"C\" \"A\" \"C\" \"C\" \"A\" \"C\" \"C\" \"A\" \"C\" \"T\" \"C\" \"T\" \"C\"\n",
       " [541] \"C\" \"T\" \"C\" \"G\" \"A\" \"C\" \"C\" \"A\" \"A\" \"G\" \"A\" \"C\" \"G\" \"A\" \"A\" \"C\" \"A\" \"G\"\n",
       " [559] \"G\" \"T\" \"G\" \"G\" \"A\" \"T\" \"G\" \"G\" \"G\" \"G\" \"A\" \"A\" \"A\" \"G\" \"G\" \"T\" \"G\" \"G\"\n",
       " [577] \"T\" \"C\" \"T\" \"T\" \"G\" \"G\" \"G\" \"C\" \"A\" \"T\" \"G\" \"C\" \"C\" \"T\" \"T\" \"C\" \"C\" \"G\"\n",
       " [595] \"C\" \"A\" \"A\" \"A\" \"A\" \"A\" \"T\" \"G\" \"C\" \"T\" \"G\" \"C\" \"C\" \"G\" \"G\" \"T\" \"G\" \"C\"\n",
       " [613] \"T\" \"T\" \"T\" \"T\" \"G\" \"G\" \"A\" \"C\" \"C\" \"T\" \"G\" \"A\" \"C\" \"C\" \"T\" \"T\" \"G\" \"C\"\n",
       " [631] \"G\" \"C\" \"G\" \"C\" \"A\" \"A\" \"G\" \"C\" \"T\" \"C\" \"T\" \"C\" \"C\" \"C\" \"A\" \"G\" \"C\" \"T\"\n",
       " [649] \"G\" \"G\" \"T\" \"A\" \"A\" \"A\" \"G\" \"A\" \"T\" \"G\" \"G\" \"A\" \"A\" \"A\" \"A\" \"G\" \"A\" \"A\"\n",
       " [667] \"C\" \"G\" \"T\" \"C\" \"A\" \"T\" \"G\" \"C\" \"G\" \"C\" \"A\" \"G\" \"C\" \"A\" \"T\" \"G\" \"G\" \"A\"\n",
       " [685] \"A\" \"T\" \"T\" \"G\" \"G\" \"T\" \"C\" \"T\" \"C\" \"T\" \"C\" \"G\" \"G\" \"G\" \"A\" \"G\" \"C\" \"G\"\n",
       " [703] \"T\" \"A\" \"T\" \"G\" \"G\" \"A\" \"G\" \"A\" \"C\" \"T\" \"G\" \"C\" \"G\" \"C\" \"A\" \"A\" \"C\" \"A\"\n",
       " [721] \"A\" \"C\" \"T\" \"G\" \"T\" \"C\" \"C\" \"A\" \"T\" \"C\" \"T\" \"G\" \"G\" \"G\" \"G\" \"T\" \"G\" \"A\"\n",
       " [739] \"G\" \"A\" \"A\" \"C\" \"T\" \"G\" \"T\" \"G\" \"A\" \"C\" \"G\" \"A\" \"A\" \"G\" \"A\" \"C\" \"G\" \"T\"\n",
       " [757] \"G\" \"T\" \"C\" \"T\" \"G\" \"A\" \"T\" \"A\" \"T\" \"C\" \"A\" \"C\" \"G\" \"G\" \"A\" \"T\" \"A\" \"A\"\n",
       " [775] \"A\" \"C\" \"T\" \"G\" \"G\" \"G\" \"T\" \"G\" \"T\" \"C\" \"T\" \"T\" \"G\" \"C\" \"T\" \"C\" \"T\" \"A\"\n",
       " [793] \"T\" \"G\" \"A\" \"A\" \"G\" \"T\" \"G\" \"G\" \"G\" \"A\" \"G\" \"A\" \"G\" \"C\" \"T\" \"G\" \"G\" \"A\"\n",
       " [811] \"G\" \"G\" \"A\" \"C\" \"A\" \"T\" \"G\" \"T\" \"A\" \"C\" \"G\" \"T\" \"C\" \"G\" \"A\" \"C\" \"C\" \"G\"\n",
       " [829] \"C\" \"T\" \"A\" \"T\" \"G\" \"A\" \"C\" \"C\" \"A\" \"G\" \"T\" \"A\" \"C\" \"C\" \"G\" \"T\" \"G\" \"T\"\n",
       " [847] \"C\" \"A\" \"C\" \"C\" \"A\" \"T\" \"C\" \"A\" \"A\" \"G\" \"A\" \"G\" \"T\" \"A\" \"T\" \"C\" \"C\" \"G\"\n",
       " [865] \"C\" \"A\" \"A\" \"T\" \"A\" \"T\" \"T\" \"G\" \"A\" \"G\" \"G\" \"C\" \"G\" \"T\" \"C\" \"C\" \"G\" \"T\"\n",
       " [883] \"C\" \"C\" \"A\" \"G\" \"C\" \"C\" \"A\" \"A\" \"G\" \"C\" \"C\" \"G\" \"T\" \"G\" \"A\" \"C\" \"C\" \"G\"\n",
       " [901] \"C\" \"A\" \"A\" \"G\" \"C\" \"A\" \"G\" \"A\" \"A\" \"G\" \"A\" \"T\" \"C\" \"A\" \"C\" \"C\" \"G\" \"A\"\n",
       " [919] \"T\" \"G\" \"A\" \"G\" \"A\" \"T\" \"T\" \"G\" \"C\" \"C\" \"A\" \"A\" \"G\" \"T\" \"T\" \"G\" \"A\" \"A\"\n",
       " [937] \"G\" \"T\" \"A\" \"C\" \"A\" \"A\" \"G\" \"G\" \"A\" \"C\" \"C\" \"C\" \"C\" \"A\" \"A\" \"C\" \"T\" \"C\"\n",
       " [955] \"C\" \"C\" \"C\" \"A\" \"C\" \"G\" \"C\" \"A\" \"T\" \"T\" \"G\" \"T\" \"C\" \"G\" \"T\" \"T\" \"C\" \"T\"\n",
       " [973] \"C\" \"G\" \"A\" \"G\" \"C\" \"A\" \"A\" \"G\" \"A\" \"A\" \"C\" \"T\" \"C\" \"G\" \"T\" \"C\" \"C\" \"G\"\n",
       " [991] \"C\" \"G\" \"C\" \"C\" \"G\" \"A\" \"G\" \"G\" \"C\" \"C\" \"G\" \"A\" \"G\" \"T\" \"C\" \"T\" \"C\" \"T\"\n",
       "[1009] \"G\" \"G\" \"T\" \"C\" \"G\" \"C\" \"T\" \"G\" \"A\" \"A\" \"G\" \"C\" \"C\" \"C\" \"A\" \"A\" \"C\" \"T\"\n",
       "[1027] \"T\" \"T\" \"C\" \"G\" \"A\" \"A\" \"C\" \"A\" \"T\" \"T\" \"A\" \"C\" \"C\" \"C\" \"G\" \"T\" \"G\" \"A\"\n",
       "[1045] \"G\" \"A\" \"A\" \"G\" \"C\" \"T\" \"C\" \"A\" \"A\" \"G\" \"G\" \"C\" \"T\" \"G\" \"C\" \"A\" \"T\" \"T\"\n",
       "[1063] \"C\" \"C\" \"A\" \"G\" \"T\" \"A\" \"C\" \"C\" \"A\" \"A\" \"T\" \"T\" \"C\" \"G\" \"A\" \"T\" \"G\" \"C\"\n",
       "[1081] \"T\" \"C\" \"T\" \"T\" \"C\" \"G\" \"A\" \"G\" \"A\" \"A\" \"C\" \"A\" \"T\" \"T\" \"G\" \"C\" \"G\" \"A\"\n",
       "[1099] \"G\" \"A\" \"A\" \"G\" \"G\" \"T\" \"C\" \"G\" \"C\" \"C\" \"A\" \"T\" \"C\" \"A\" \"T\" \"C\" \"G\" \"C\"\n",
       "[1117] \"C\" \"G\" \"G\" \"G\" \"T\" \"A\" \"T\" \"G\" \"G\" \"G\" \"A\" \"A\" \"A\" \"C\" \"A\" \"C\" \"C\" \"T\"\n",
       "[1135] \"T\" \"C\" \"T\" \"C\" \"G\" \"A\" \"C\" \"C\" \"T\" \"C\" \"A\" \"T\" \"C\" \"G\" \"A\" \"T\" \"G\" \"A\"\n",
       "[1153] \"C\" \"A\" \"C\" \"T\" \"C\" \"C\" \"G\" \"G\" \"T\" \"C\" \"A\" \"C\" \"T\" \"C\" \"C\" \"T\" \"G\" \"G\"\n",
       "[1171] \"T\" \"G\" \"A\" \"G\" \"A\" \"C\" \"T\" \"C\" \"G\" \"C\" \"C\" \"A\" \"G\" \"G\" \"C\" \"G\" \"T\" \"A\"\n",
       "[1189] \"T\" \"G\" \"A\" \"C\" \"G\" \"G\" \"A\" \"T\" \"A\" \"C\" \"G\" \"A\" \"C\" \"G\" \"C\" \"T\" \"A\" \"G\"\n",
       "[1207] \"C\" \"A\" \"A\" \"G\" \"G\" \"C\" \"C\" \"A\" \"T\" \"C\" \"A\" \"T\" \"T\" \"C\" \"A\" \"G\" \"G\" \"A\"\n",
       "[1225] \"C\" \"T\" \"G\" \"C\" \"G\" \"A\" \"A\" \"G\" \"A\" \"C\" \"G\" \"C\" \"C\" \"T\" \"T\" \"G\" \"G\" \"C\"\n",
       "[1243] \"A\" \"A\" \"A\" \"C\" \"T\" \"G\" \"G\" \"G\" \"T\" \"C\" \"T\" \"C\" \"G\" \"T\" \"C\" \"C\" \"A\" \"A\"\n",
       "[1261] \"G\" \"G\" \"C\" \"T\" \"G\" \"C\" \"C\" \"G\" \"T\" \"C\" \"A\" \"A\" \"A\" \"T\" \"C\" \"C\" \"A\" \"G\"\n",
       "[1279] \"C\" \"T\" \"T\" \"G\" \"T\" \"C\" \"C\" \"C\" \"A\" \"G\" \"C\" \"G\" \"T\" \"T\" \"C\" \"A\" \"C\" \"G\"\n",
       "[1297] \"T\" \"A\" \"C\" \"T\" \"C\" \"T\" \"C\" \"T\" \"C\" \"T\" \"C\" \"A\" \"G\" \"C\" \"G\" \"C\" \"C\" \"G\"\n",
       "[1315] \"A\" \"C\" \"A\" \"G\" \"G\" \"A\" \"T\" \"A\" \"A\" \"C\" \"A\" \"T\" \"C\" \"A\" \"A\" \"T\" \"A\" \"A\"\n",
       "[1333] \"G\" \"G\" \"G\" \"C\" \"C\" \"G\" \"T\" \"G\" \"A\" \"A\" \"A\" \"A\" \"T\" \"G\" \"G\" \"G\" \"G\" \"T\"\n",
       "[1351] \"G\" \"G\" \"A\" \"C\" \"C\" \"T\" \"G\" \"T\" \"C\" \"G\" \"T\" \"C\" \"C\" \"C\" \"A\" \"G\" \"G\" \"A\"\n",
       "[1369] \"C\" \"C\" \"A\" \"G\" \"C\" \"C\" \"C\" \"T\" \"T\" \"G\" \"A\" \"G\" \"G\" \"G\" \"G\" \"T\" \"G\" \"A\"\n",
       "[1387] \"C\" \"C\" \"G\" \"G\" \"G\" \"A\" \"T\" \"T\" \"C\" \"A\" \"T\" \"G\" \"G\" \"G\" \"T\" \"T\" \"G\" \"C\"\n",
       "[1405] \"C\" \"G\" \"C\" \"A\" \"G\" \"A\" \"G\" \"C\" \"A\" \"G\" \"C\" \"A\" \"C\" \"C\" \"C\" \"T\" \"A\" \"G\"\n",
       "[1423] \"C\" \"T\" \"A\" \"C\" \"C\" \"A\" \"C\" \"C\" \"A\" \"T\" \"G\" \"A\" \"C\" \"G\" \"A\" \"T\" \"G\" \"A\"\n",
       "[1441] \"T\" \"G\" \"A\" \"A\" \"G\" \"A\" \"G\" \"G\" \"C\" \"T\" \"G\" \"C\" \"C\" \"A\" \"G\" \"T\" \"G\" \"A\"\n",
       "[1459] \"T\" \"G\" \"G\" \"A\" \"G\" \"A\" \"C\" \"A\" \"T\" \"C\" \"C\" \"G\" \"C\" \"G\" \"G\" \"C\" \"C\" \"G\"\n",
       "[1477] \"C\" \"G\" \"A\" \"A\" \"G\" \"A\" \"A\" \"G\" \"A\" \"A\" \"C\" \"G\" \"G\" \"G\" \"A\" \"G\" \"C\" \"C\"\n",
       "[1495] \"T\" \"G\" \"T\" \"C\" \"C\" \"C\" \"T\" \"G\" \"T\" \"C\" \"T\" \"A\" \"A\" \"T\" \"C\" \"G\" \"A\" \"G\"\n",
       "[1513] \"C\" \"A\" \"C\" \"T\" \"C\" \"A\" \"C\" \"T\" \"G\" \"A\" \"A\" \"T\" \"G\" \"A\" \"G\" \"C\" \"A\" \"G\"\n",
       "[1531] \"G\" \"G\" \"C\" \"G\" \"C\" \"A\" \"A\" \"G\" \"A\" \"A\" \"T\" \"A\" \"C\" \"C\" \"G\" \"G\" \"C\" \"A\"\n",
       "[1549] \"A\" \"A\" \"A\" \"C\" \"G\" \"A\" \"T\" \"A\" \"C\" \"A\" \"T\" \"T\" \"G\" \"G\" \"T\" \"T\" \"A\" \"A\"\n",
       "[1567] \"T\" \"C\" \"A\" \"C\" \"T\" \"T\" \"T\" \"G\" \"A\" \"T\" \"G\" \"C\" \"T\" \"C\" \"T\" \"T\" \"G\" \"G\"\n",
       "[1585] \"A\" \"C\" \"T\" \"C\" \"C\" \"A\" \"T\" \"A\" \"A\" \"T\" \"T\" \"A\" \"T\" \"T\" \"G\" \"T\" \"A\" \"C\"\n",
       "[1603] \"G\" \"A\" \"A\" \"T\" \"G\" \"A\" \"C\" \"T\" \"G\" \"T\" \"T\" \"T\" \"T\" \"C\" \"C\" \"C\" \"G\" \"C\"\n",
       "[1621] \"T\" \"T\" \"G\" \"A\" \"T\" \"C\" \"T\" \"T\" \"C\" \"A\" \"G\" \"G\" \"A\" \"T\" \"T\" \"C\" \"A\" \"G\"\n",
       "[1639] \"A\" \"T\" \"C\" \"G\" \"G\" \"T\" \"G\" \"T\" \"C\" \"T\" \"G\" \"C\" \"C\" \"A\" \"C\" \"G\" \"T\" \"T\"\n",
       "[1657] \"A\" \"T\" \"G\" \"A\" \"T\" \"G\" \"T\" \"T\" \"A\" \"C\" \"T\" \"T\" \"A\" \"T\" \"T\" \"A\" \"C\" \"C\"\n",
       "[1675] \"C\" \"C\" \"T\" \"T\" \"C\" \"G\" \"G\" \"T\" \"T\" \"T\" \"G\" \"A\" \"T\" \"A\" \"A\" \"T\" \"G\" \"C\"\n",
       "[1693] \"T\" \"T\" \"A\" \"C\" \"T\" \"T\" \"G\" \"C\" \"T\" \"T\" \"C\" \"T\" \"G\" \"G\" \"C\" \"A\" \"A\" \"T\"\n",
       "[1711] \"A\" \"A\" \"T\" \"C\" \"A\" \"A\" \"C\" \"C\" \"A\" \"A\" \"T\" \"T\" \"C\" \"A\" \"A\" \"A\" \"A\" \"C\"\n",
       "[1729] \"G\" \"C\" \"T\" \"T\" \"C\" \"C\" \"C\" \"C\" \"A\" \"A\" \"T\" \"T\" \"G\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Split the cleaned sequence into individual letters\n",
    "pilA_vector_split <- stringr::str_split( pilA_vector ,pattern = \"\" ,simplify = FALSE)[[1]] # The [[1]] at the end retrieves the first element of the resulting list, which contains the split sequence as a character vector.\n",
    "pilA_vector_split\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0391eb15",
   "metadata": {},
   "source": [
    "## Length and Base composition of a DNA sequence  \n",
    "\n",
    "\n",
    "<div style=\"text-align: justify\">Once you have retrieved a DNA sequence, we can obtain some simple statistics to describe that sequence, such as the sequence’s total length in nucleotides. In the above example, we retrieved the pilA sequence, and stored it in the vector variable pilA_vector_split. To obtain the length of the genome sequence, we would use the `length()` function.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "cd642d7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1741"
      ],
      "text/latex": [
       "1741"
      ],
      "text/markdown": [
       "1741"
      ],
      "text/plain": [
       "[1] 1741"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Return the length of the pilA_vector_split vector\n",
    "length (pilA_vector_split)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6883b737",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\">Every analysis of a DNA sequence often involves counting the occurrences of the four nucleotides: \"A,\" \"C,\" \"G,\" and \"T.\" This can be accomplished using the table() function. For instance, to determine the counts of A, C, G, and T in the pilA sequence (stored in the vector variable pilA_vector_split  as per the previous commands), you can use the following code:</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "57e11164",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pilA_vector_split\n",
       "  A   C   G   T \n",
       "403 502 426 410 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Display  counts of each nucleotide in the sequence\n",
    "table(pilA_vector_split)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "52e46355",
   "metadata": {},
   "source": [
    "## GC Content of DNA \n",
    "\n",
    "<div style=\"text-align: justify\">One of the fundamental properties of a genome sequence is its GC content, which represents the percentage of Gs and Cs in the sequence. To calculate the GC content, we need to determine the number of Gs and Cs in the sequence and divide it by the total length of the genome.</div>\n",
    "\n",
    "<div style=\"text-align: justify\">For example, let's consider the pilA sequence. Using the table() function as mentioned earlier, we found that it contains 403 As, 502 Cs, 436 Gs, and 410 Ts. To calculate the GC content, you can use the following formula:  </div>\n",
    "\n",
    "GC content = (number of Gs + number of Cs) / (genome length) * 100\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "11a09466",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "53.877082136703"
      ],
      "text/latex": [
       "53.877082136703"
      ],
      "text/markdown": [
       "53.877082136703"
      ],
      "text/plain": [
       "[1] 53.87708"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Manual calculation of GC content m\n",
    "GC_content = ((502+436)/1741)*100\n",
    "GC_content"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23adce40",
   "metadata": {},
   "source": [
    "Alternatively, we can use the GC() function in the SeqinR package, which simplifies the calculation by directly providing the fraction of Gs and Cs in the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "7d4de8d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "53.3026995979322"
      ],
      "text/latex": [
       "53.3026995979322"
      ],
      "text/markdown": [
       "53.3026995979322"
      ],
      "text/plain": [
       "[1] 53.3027"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use of the function from the SeqinR package to calculate the GC content of a DNA sequence\n",
    "seqinr::GC(pilA_vector_split) *100"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a3aa6567",
   "metadata": {},
   "source": [
    "## Finding patterns/motifs \n",
    "<div style=\"text-align: justify\">In the field of biology, finding patterns is a common task that involves identifying specific sequences or motifs within biological data such as DNA, RNA, or protein sequences. These patterns can provide valuable insights into the structure, function, and evolution of biological molecules.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "156e30ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>99</li><li>651</li><li>772</li><li>1321</li><li>1330</li><li>1505</li><li>1564</li><li>1591</li><li>1687</li><li>1710</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 99\n",
       "\\item 651\n",
       "\\item 772\n",
       "\\item 1321\n",
       "\\item 1330\n",
       "\\item 1505\n",
       "\\item 1564\n",
       "\\item 1591\n",
       "\\item 1687\n",
       "\\item 1710\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 99\n",
       "2. 651\n",
       "3. 772\n",
       "4. 1321\n",
       "5. 1330\n",
       "6. 1505\n",
       "7. 1564\n",
       "8. 1591\n",
       "9. 1687\n",
       "10. 1710\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1]   99  651  772 1321 1330 1505 1564 1591 1687 1710"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 10\n"
     ]
    }
   ],
   "source": [
    "# Find the positions of a given motif within a sequence\n",
    "motif <- \"TAA\"\n",
    "words.pos(motif, pilA_vector, ignore.case = FALSE,perl = TRUE) #pilA_vector is a single line character vector \n",
    "# Print their frequency \n",
    "print(length(words.pos(motif, pilA_vector, ignore.case = FALSE,perl = TRUE)))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9809a67b",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\">In conclusion, sequences play a critical role in biological research, serving as fundamental tools for analyzing genetic information. Through sequence analysis, researchers can gain valuable insights into various aspects of genetics and genomics. In this tutorial, we have covered essential topics related to sequence analysis using R. We have learned how to read and store sequences, manipulate and transform them, find motifs, and calculate sequence statistics."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03d7f25b",
   "metadata": {},
   "source": [
    "## References  \n",
    "\n",
    "\n",
    "1.\tOrtutay, C. & Ortutay, Z. Molecular Data Analysis Using R. (Wiley-Blackwell, 2017).  \n",
    "\n",
    "2.\tVangelatos, I. et al. Eisosome Organization in the Filamentous AscomyceteAspergillus nidulans. Eukaryot. Cell 9, 1441–1454 (2010).  \n",
    "\n",
    "3.\tAthanasopoulos, A., Gournas, C., Amillis, S. & Sophianopoulou, V. Characterization of AnNce102 and its role in eisosome stability and sphingolipid biosynthesis. Sci. Rep. 5, (2015).  \n",
    "\n",
    "4.\tAthanasopoulos, A., André, B., Sophianopoulou, V. & Gournas, C. Fungal plasma membrane domains. FEMS Microbiol. Rev. fuz022 (2019) doi:10.1093/femsre/fuz022.  \n",
    "\n",
    "5.\tBrouwer, A. C., with contributions by Nathan L. Chapter 21 Downloading DNA sequences as FASTA files in R | A Little Book of R for Bioinformatics 2.0.  \n",
    "\n",
    "6.\tBrouwer, A. C., with contributions by Nathan L. A Little Book of R for Bioinformatics 2.0.  \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f39f6d1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
